STAGING - 21.0.0.10 GA Release Blog Post (#1753)
OpenLiberty/blogs: posts/2021-10-01-ga-release-post.adoc

In link:{url-about}[Open Liberty] 21.0.0.10:

* <<java17, Java 17 support>>

View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A210010+label%3A%22release+bug%22

[#run]
== Run your apps using 21.0.0.10

If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>21.0.0.10</version>
    <type>zip</type>
</dependency>
----

Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

[source,gradle]
----
dependencies {
    libertyRuntime group: 'io.openliberty', name: 'openliberty-runtime', version: '[21.0.0.10,)'
}
----

Or if you're using Docker:

[source]
----
FROM open-liberty
----

Or take a look at our link:{url-prefix}/downloads/[Downloads page].

[link=https://stackoverflow.com/tags/open-liberty]
image::img/blog/blog_btn_stack.svg[Ask a question on Stack Overflow, align="center"]

[#java17]

> === Java 17 Support

Reviewer: ```suggestion
== Java 17 Support
```

* link:https://openjdk.java.net/jeps/358[Helpful NullPointerExceptions]
* link:https://openjdk.java.net/jeps/361[Switch Expressions]
* link:https://openjdk.java.net/jeps/372[Removal of the Nashorn JavaScript Engine]
* link:https://openjdk.java.net/jeps/377[ZGC: A Scalable Low-Latency Garbage Collector]
* link:https://openjdk.java.net/jeps/378[Text Blocks]
* link:https://openjdk.java.net/jeps/380[Unix-Domain Socket Channels]
* link:https://openjdk.java.net/jeps/390[Warnings for Value-Based Classes]
* link:https://jdk.java.net/16/release-notes#JDK-8202343[Removal of TLS 1.0 and 1.1]
* link:https://openjdk.java.net/jeps/394[Pattern Matching for instanceof]
* link:https://openjdk.java.net/jeps/395[Records]
* link:https://openjdk.java.net/jeps/403[Strongly Encapsulate JDK Internals]
* link:https://openjdk.java.net/jeps/409[Sealed Classes]
* link:https://openjdk.java.net/jeps/411[Deprecate the Security Manager for Removal]
* link:https://bugs.openjdk.java.net/browse/JDK-8196415[Disable SHA-1 JARs]
* link:https://bugs.openjdk.java.net/browse/JDK-8259709[Disable SHA-1 XML Signatures]

> == Phasing out the --illegal-access JVM option

Reviewer: ```suggestion
=== Phasing out the --illegal-access JVM option
```

By far one of the the biggest impacts to Java development happened over the past couple releases. Java introduced a key play to move Java applications firmly into the modular programming world of the Java Platform Module System (JPMS).

> For a quick history review of JPMS, please re-visit the *Introduction of the Java Platform Module System (JPMS)* section of our link:https://openliberty.io/blog/2019/02/06/java-11.html[Open Liberty and Java 11 blog].

Reviewer: ```suggestion
For a quick history review of JPMS, please re-visit the *Introduction of the Java Platform Module System (JPMS)* section of our link:{url-prefix}/blog/2019/02/06/java-11.html[Open Liberty and Java 11 blog].
```

```
OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=permit; support was removed in 17.0
```

Not to worry though.  You still have more time if your applications or dependencies are not completely ready to run in the modularized Java world.  There are some more precise JVM arguments, primarily `--add-opens` and `--add-exports`, that can help bridge the gap until you, and your applications, are fully modular ready.  These arguments allow the individual module exceptions to the JPMS rules that your applications might need.  

For more information on this topic, please review the link:https://docs.oracle.com/en/java/javase/17/migrate/migrating-jdk-8-later-jdk-releases.html#GUID-7BB28E4D-99B3-4078-BDC4-FC24180CE82B[Strong Encapsulation in the JDK] documentation or checkout one of the other helpful links in the link:<Please turn this into a link to the Java 17 resources section>[Java 17 resources] sections below.

> == How to run Open Liberty on Java 17

Reviewer: ```suggestion
=== How to run Open Liberty on Java 17
```

You will need a copy of Java 17.  To try this out, download Java 17 Temurin from link:https://adoptium.net/?variant=openjdk17[adoptium.net].

Then you will need Open Liberty 21.0.0.10, or newer. To run your applications on Open Liberty 21.0.0.10, check out <<run, Run your apps using 21.0.0.10>>.

There are a number of ways to control the Java executable used by the Open Liberty server. The following list describes the order of precedence:

. `JAVA_HOME` in `${server.config.dir}/server.env`
. `JAVA_HOME` in `${wlp.install.dir}/etc/server.env`
. `JAVA_HOME` set in the environment
. searching the `PATH` in the environment

 For more information, please visit our link:{url-prefix}/docs/latest/reference/config/server-configuration-overview.html[server configuration documentation].

> Happy Java 17 experimenting!

Reviewer 2: This section is not showing up correctly.  The link is not a link and the whole thing is in a code block, but should not be.

> == Java 17 resources

Reviewer: ```suggestion
=== Java 17 resources
```

For further reading on Java 17, here are a few resources you might find useful:

> * Highlights of new features in Java 17 are found in the link:https://openjdk.java.net/projects/jdk/17/[Java 17 features page]

Reviewer 3: I believe this was suposed to be a bulleted list, but it is not rendering as such.

Reviewer 3: Still isn't rendered correctly from what I can see.

Reviewer 2: You are correct Jared, these are supposed to be bulleted lists and they are not showing up correctly for me either.

Reviewer 4: iirc, there needs to be a blank line above the start of a bulleted list in asciidoc - maybe try:
```suggestion

* Highlights of new features in Java 17 are found in the link:https://openjdk.java.net/projects/jdk/17/[Java 17 features page]
```

For migration to Java 17, please check out link:https://blogs.oracle.com/javamagazine/migrate-to-java-17[Oracle Java 17 migration blog] and link:https://docs.oracle.com/en/java/javase/17/migrate/getting-started.html[Getting started with JDK 17].

For Java 17 release roadmaps and availability refer to the link:https://adoptium.net/support.html?variant=openjdk17&jvmVariant=hotspot[Adoptium release roadmap] and link:https://www.oracle.com/java/technologies/java-se-support-roadmap.html[Oracle commercial JDK roadmap].

And finally, for a comprehensive discussion on Java modules with a little hands on practice, read link:https://www.baeldung.com/java-9-modularity[this article from Baeldung].

[#bugs]
== Notable bugs fixed in this release

We’ve spent some time fixing bugs. The following sections describe just some of the issues resolved in this release. If you’re interested, here’s the  link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A210010+label%3A%22release+bug%22[full list of bugs fixed in 21.0.0.10].

* link:https://github.com/OpenLiberty/open-liberty/issues/18393[SPNEGO does not work with Java 11 or higher]
+

> In the release of Open Liberty 21.0.0.9, a regression was introduced which prevented the `Krb5LoginModule` class from being found when authenticating using Kerberos on Java 11 or higher. This issue caused some users to be presented with an `Error 403` or `Authentication Failed` when authenticating using the `spnego-1.0` feature. This issue has been fixed by reverting the underlying code back to the way it was prior to it being modified. 

Reviewer 3: This issue was NOT fixed by reverting the underlying code.  The last sentence should say something like `This issue was fixed by changing how classes are loaded to work with Java 11 and hgiher releases.` or something like that.

* link:https://github.com/OpenLiberty/open-liberty/issues/18559[Liberty is vulnerable to Information Disclosure (CVE-2021-29842 CVSS 3.7)]
+
In the 21.0.0.9 release, an Information Disclosure vulnerability was discovered which could cause sensitive information to unintentionally be revealed. This issue has now been fixed in Open Liberty 21.0.0.10.

* link:https://github.com/OpenLiberty/open-liberty/issues/18475[Servlet ReadListener does not receive all HTTP request data]
+

> In 21.0.0.9 an issue was discovered where in rare cases, an app's `javax.servlet.ReadListener` could be invoked with `onAllDataRead()` prematurely, before all of the bytes sent along with the request have been made available to the app. In this scenario, `onDataAvailable()` would have been incorrectly invoked after the premature `onAllDataRead()` to provide notification for any remaining request data. This issue has been fixed by setting the `HttpServiceContextImpl` body to complete later on in the code.

Reviewer 5: I don't think we should start this with `In 21.0.0.9`, that implies to me that this wasn't an issue in earlier releases. Suggested edit:

An issue was discovered where in rare cases an app's `javax.servlet.ReadListener` could be invoked with `onAllDataRead()` prematurely, before all of the bytes sent along with the request have been made available to the `ReadListener`. In an affected scenario, `onDataAvailable()` would be incorrectly invoked after the premature `onAllDataRead()` to provide notification for any remaining request data. This issue has been fixed with an update to the HTTP implementation which prevents HTTP request bodies from being prematurely marked as complete.

* link:https://github.com/OpenLiberty/open-liberty/issues/18552[JAX-RS 2.0 and 2.1 implementation is executing resource method when Content-Type or Accept header contains invalid values]
+

> An issue was discovered in 21.0.0.9 which meant that JAX-RS (2.0 and 2.1) apps would execute resource code when bad requests were made using invalid `Content-Type` or `Accept` headers, rather than returning a 400 (Bad Request). A fix has now been implemented to ensure that if a request is made with invalid `Content-Type` or `Accept` headers, the application returns 400 (Bad Request).

Reviewer 4: ```suggestion
An issue was discovered in 21.0.0.9 which meant that Liberty's JAX-RS (2.0 and 2.1) implementation would execute resource code when bad requests were made using invalid `Content-Type` or `Accept` headers, rather than returning a 400 (Bad Request). A fix has now been implemented to ensure that if a request is made with invalid `Content-Type` or `Accept` headers, the implementation returns a 400 (Bad Request).
```

Substituting "implementation" for "application" to indicate that Liberty is doing the work of returning the 400 as opposed to the application.

* link:https://github.com/OpenLiberty/open-liberty/issues/18348[ContainerRequestContext.getAcceptableLanguages() - fails with IllegalArgumentException when invalid locales are specified in the Accept-Language header]
+
A bug was discovered which meant that `ContainerRequestContext.getAcceptableLanguages()` would throw an `IllegalArgumentException` when requests were made with invalid locales specified in the `Accept-Language` header. This `getAcceptableLanguages()` call should succeed and return the list from the headers as submitted. This issue has now been fixed to ensure that the `Accept-Language` header is handled consistently.

* link:https://github.com/OpenLiberty/open-liberty/issues/18437[JSF throws ClassNotFoundException for o.a.m.el.convert.ValueExpressionToValueBinding]
+

> Previously, A `JSF` application using `jsf-2.2` or `jsf-2.3` could encounter a `ClassNotFoundException` for the internal `MyFaces` class `ValueExpressionToValueBinding`, even if that class wasn't directly referenced by the app. This issue occured because the `org.apache.myfaces.el.convert` package was not made available (by Liberty's `jsf-2.2` and `jsf-2.3` bundles) to the thread context `classloader`. This visibility issue was fixed by updating those bundles.

Reviewer 5: I'd like to add a bit more information (which I've only just now added to OL issue #18437:

Previously, A JSF application using [Application.createValueBinding(String)](https://docs.oracle.com/javaee/6/api/javax/faces/application/Application.html#createValueBinding(java.lang.String)) and `jsf-2.2` or `jsf-2.3` could encounter a `ClassNotFoundException` for the internal `MyFaces` class `ValueExpressionToValueBinding` ...

* link:https://github.com/OpenLiberty/open-liberty/issues/18503[RuntimeCodebase cannot be located on collocated call]
+

> When invoking a Remote EJB in the same server, the call will be routed internally (collocated) rather than via the network. If the calling app, and the called app have different versions of the same class, negotiation is required between them via the `RuntimeCodebase` object. An issue was discovered where the `RuntimeCodebase` was not available for collocated calls. This issue was fixed in Open Liberty 21.0.0.10, meaning that `Yoko` now supports these collocated calls.

Reviewer 6: Change the opening to 

> When an app invokes a remote enterprise bean in the same Liberty server

This removes the dangling participle and uses the more modern terminology that includes Java EE and Jakarta EE.

Reviewer: ```suggestion
When an app invokes a remote enterprise bean in the same Liberty server, the call will be routed internally (collocated) rather than via the network. If the calling app, and the called app have different versions of the same class, negotiation is required between them via the `RuntimeCodebase` object. An issue was discovered where the `RuntimeCodebase` was not available for collocated calls. This issue was fixed in Open Liberty 21.0.0.10, meaning that `Yoko` now supports these collocated calls.
```

author_github: https://github.com/ryan-storey
seo-title: Java 17 support and a number of significant bug fixes in Open Liberty 21.0.0.10 - OpenLiberty.io
seo-description: The release of Open Liberty 21.0.0.10 offers the long-awaited Java 17 support and its many new enhancements and functionality!
blog_description: The release of Open Liberty 21.0.0.10 offers the long-awaited Java 17 support and its many new enhancements and functionality!
open-graph-image: https://openliberty.io/img/twitter_card.jpg
---
= Java 17 support and a number of significant bug fixes in Open Liberty 21.0.0.10
Ryan Storey <https://github.com/ryan-storey>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

The release of Open Liberty 21.0.0.10 offers the long-awaited Java 17 support and its many new enhancements and functionality! This release also provides several important new bug fixes.

In link:{url-about}[Open Liberty] 21.0.0.10:

* <<java17, Java 17 support>>

View the list of fixed bugs in link:https://github.com/OpenLiberty/open-liberty/issues?q=label%3Arelease%3A210010+label%3A%22release+bug%22

[#run]
== Run your apps using 21.0.0.10

> If you're using link:{url-prefix}/guides/maven-intro.html[Maven], here are the coordinates:

Reviewer 2: Maven link does not work in staging

Author: This is expected behaviour don't worry, as the {url-prefix} is the hostname, i.e. https://openliberty.io, or in this case,  https://blogs-draft-openlibertyio.mybluemix.net. The draft site only works for blogs and therefore does not have the maven guide, whereas on the prod site the {url-prefix} will be https://openliberty.io and therefore will work.

[source,xml]
----
<dependency>
    <groupId>io.openliberty</groupId>
    <artifactId>openliberty-runtime</artifactId>
    <version>21.0.0.10</version>
    <type>zip</type>
</dependency>
----

> Or for link:{url-prefix}/guides/gradle-intro.html[Gradle]:

Reviewer 2: Gradle link does not work in staging

Then, in Java 17 with link:https://openjdk.java.net/jeps/403[JEP 403], they took strong encapsulation further.  Now, while the JVM option `--illegal-access` is still recognized as a valid JVM argument, it is effectively hard coded to `deny`, its passed in value is ignored and the following warning message is issued.

```
OpenJDK 64-Bit Server VM warning: Ignoring option --illegal-access=permit; support was removed in 17.0
```

Not to worry though.  You still have more time if your applications or dependencies are not completely ready to run in the modularized Java world.  There are some more precise JVM arguments, primarily `--add-opens` and `--add-exports`, that can help bridge the gap until you, and your applications, are fully modular ready.  These arguments allow the individual module exceptions to the JPMS rules that your applications might need.  

> For more information on this topic, please review the link:https://docs.oracle.com/en/java/javase/17/migrate/migrating-jdk-8-later-jdk-releases.html#GUID-7BB28E4D-99B3-4078-BDC4-FC24180CE82B[Strong Encapsulation in the JDK] documentation or checkout one of the other helpful links in the link:<Please turn this into a link to the Java 17 resources section>[Java 17 resources] sections below.

Reviewer 2: Just to make sure this does not get missed, the text `link:<Please turn this into a link to the Java 17 resources section>[Java 17 resources]` will need to be changed into a valid link.  Looks like the whole text should be changed to `<<java-17-resources, Java 17 resources>>`.

Author: Thanks for pointing that out!

Reviewer 2: I unresolved this conversation because this psudo link is still showing on the staging server.

=== How to run Open Liberty on Java 17

You will need a copy of Java 17.  To try this out, download Java 17 Temurin from link:https://adoptium.net/?variant=openjdk17[adoptium.net].

Then you will need Open Liberty 21.0.0.10, or newer. To run your applications on Open Liberty 21.0.0.10, check out <<run, Run your apps using 21.0.0.10>>.

There are a number of ways to control the Java executable used by the Open Liberty server. The following list describes the order of precedence:

. `JAVA_HOME` in `${server.config.dir}/server.env`
. `JAVA_HOME` in `${wlp.install.dir}/etc/server.env`
. `JAVA_HOME` set in the environment
. searching the `PATH` in the environment

> For more information, please visit our link:{url-prefix}/docs/latest/reference/config/server-configuration-overview.html[server configuration documentation].

Reviewer 2: Server config link does not work in staging
